File: scripts/generate_index.py
#!/usr/bin/env python3
"""Generate index.json for the repository — file listing with sizes."""

import json, os, sys
from pathlib import Path

REPO_ROOT = Path(__file__).parent.parent.resolve()
OUT_FILE = REPO_ROOT / "index.json"
IGNORE = {".git", ".github", "index.html", "index.json", "generate_index.py", "README.md", ".gitignore"}

def files_json(path=REPO_ROOT, prefix=""):
    entries = []
    for entry in sorted(path.iterdir()):
        name = entry.name
        if name in IGNORE or name.startswith("."):
            continue
        rel = (prefix + name) if prefix else name
        if entry.is_dir():
            entries.append({"name": rel + "/", "type": "dir", "size": None})
            entries.extend(files_json(entry, rel + "/"))
        else:
            entries.append({"name": rel, "type": "file", "size": entry.stat().st_size})
    return entries

def main():
    entries = files_json()
    with open(OUT_FILE, "w") as f:
        json.dump(entries, f, indent=2)
    print(f"index.json written ({len(entries)} entries)", file=sys.stderr)

if __name__ == "__main__":
    main()

File: src/_scripts/generate_index.mjs
#!/usr/bin/env node
/** Generate index.json — file listing with sizes, by category. */

import { writeFileSync, readdirSync, statSync } from "fs";
import { join, extname } from "path";
import { fileURLToPath } from "url";

const __dirname = fileURLToPath(new URL(".", import.meta.url));
const REPO_ROOT = join(__dirname, "../..");
const DEFAULT_OUT_FILE = join(REPO_ROOT, "index.json");

const outFlagIndex = process.argv.indexOf("--out");
const OUT_FILE = outFlagIndex >= 0 && process.argv[outFlagIndex + 1]
  ? process.argv[outFlagIndex + 1]
  : DEFAULT_OUT_FILE;

const IGNORE = new Set([
  ".git",
  ".github",
  "index.html",
  "index.json",
  "generate_index.mjs",
  "README.md",
  ".gitignore",
  "_scripts",
  "_templates",
  "node_modules",
]);

/**
 * Keep folder-based categories stable:
 * - src/documents/<group>/file.html -> <group>
 * - src/images/file.png -> images
 */
function inferCategory(relPath) {
  const parts = relPath.split("/");
  const topLevel = parts[1] || "documents";
  if (topLevel === "documents" && parts.length > 3) {
    return parts[2];
  }
  return topLevel;
}

/** @param {string} dir @param {string} [prefix] @returns {any[]} */
function walk(dir, prefix = "") {
  const entries = [];
  for (const entry of readdirSync(dir).sort()) {
    if (IGNORE.has(entry) || entry.startsWith(".")) continue;
    const full = join(dir, entry);
    const rel = prefix ? `${prefix}/${entry}` : entry;
    const stat = statSync(full);
    if (stat.isDirectory()) {
      entries.push(...walk(full, rel));
    } else {
      const ext = extname(entry).replace(".", "").toLowerCase() || "txt";
      entries.push({
        name: entry.replace(/\.[^.]+$/, ""),
        path: rel,
        category: inferCategory(rel),
        ext,
        size: stat.size,
      });
    }
  }
  return entries;
}

const files = walk(REPO_ROOT);
writeFileSync(OUT_FILE, JSON.stringify(files, null, 2));
console.error(`index.json written (${files.length} entries)`);
